openquake/commonlib/readinput.py

```diff
@@ -472,46 +472,41 @@ def get_source_models(oqparam, source_model_lt, sitecol=None, in_memory=True):
             sm, weight, smpath, trt_models, gsim_lt, i, num_samples, None)
 
 
-def get_composite_source_model(
-        oqparam, sitecol=None, SourceProcessor=source.SourceFilterSplitter,
-        monitor=DummyMonitor(), no_distribute=parallel.no_distribute(),
-        dstore=Fake()):
+def get_composite_source_model(oqparam, in_memory=True):
     """
-    Build the source models by splitting the sources. If prefiltering is
-    enabled, also reduce the GSIM logic trees in the underlying source models.
+    Parse the XML and build a complete composite source model in memory.
 
     :param oqparam:
         an :class:`openquake.commonlib.oqvalidation.OqParam` instance
-    :param sitecol:
-        a :class:`openquake.hazardlib.site.SiteCollection` instance
-    :param SourceProcessor:
-        the class used to process the sources
-    :param monitor:
-        a monitor instance
-    :param no_distribute:
-        used to disable parallel splitting of the sources
-    :param dstore:
-        a DataStore instance (possibly fake)
-    :returns:
-        an iterator over :class:`openquake.commonlib.source.SourceModel`
+    :param in_memory:
+        if False, just parse the XML without instanting the sources
     """
-    processor = SourceProcessor(sitecol, oqparam.maximum_distance, monitor)
     source_model_lt = get_source_model_lt(oqparam)
     smodels = []
     trt_id = 0
+    idx = 0
+
+    def getid(src):
+        try:
+            return src.source_id
+        except:
+            return src['id']
     for source_model in get_source_models(
-            oqparam, source_model_lt, processor.sitecol,
-            in_memory=hasattr(processor, 'process')):
+            oqparam, source_model_lt, in_memory=in_memory):
         for trt_model in source_model.trt_models:
+            trt_model.sources = sorted(trt_model, key=getid)
             trt_model.id = trt_id
+            for src in trt_model:
+                if hasattr(src, 'trt_model_id'):
+                    # .trt_model_id is missing for source nodes
```

Comment:
Please further explain why in some cases we need to enumerate, and in others we don't